sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala

```diff
@@ -0,0 +1,174 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.shuffle
+
+import scala.collection.mutable.ArrayBuffer
+
+// Helper trait that callers can use to add blocks to the iterator
+// as long as they can provide a size
+trait BlockWithSize {
+  /**
+   * Abstract method to return the size in bytes of this block
+   * @return Long - size in bytes
+   */
+  def size: Long
+}
+
+/**
+ * Specifies a start and end range of bytes for a block.
+ * @param block - a BlockWithSize instance
+ * @param rangeStart - byte offset for the start of the range
+ * @param rangeEnd - byte offset for the end of the range
+ * @tparam T - the specific type of `BlockWithSize`
+ */
+case class BlockRange[T <: BlockWithSize](
+    block: T, rangeStart: Long, rangeEnd: Long) {
+  require(rangeStart <= rangeEnd,
+    s"Instantiated a BlockRange with invalid boundaries: $rangeStart to $rangeEnd")
+
+  /**
+   * Returns the size of this range in bytes
+   * @return - Long - size in bytes
+   */
+  def rangeSize(): Long = rangeEnd - rangeStart + 1
+
+  def isComplete(): Boolean = rangeEnd == block.size - 1
+}
+
+/**
+ * Given a set of blocks, this iterator returns BlockRanges
+ * of such blocks that fit `windowSize`. The ranges are just logical
+ * chunks of the blocks, so this class performs no memory management or copying.
+ *
+ * If a block is too large for the window, the block will be
+ * returned in `next()` until the full block can be covered.
+ *
+ * For example, given a block that is 4 window-sizes in length:
+ * block = [sb1, sb2, sb3, sb4]
+ *
+ * The window will return on `next()` four "sub-blocks", governed by `windowSize`:
+ * window.next() // sb1
+ * window.next() // sb2
+ * window.next() // sb3
+ * window.next() // sb4
+ *
+ * If blocks are smaller than the `windowSize`, they will be packed:
+ * block1 = [b1]
+ * block2 = [b2]
+ * window.next() // [b1, b2]
+ *
+ * A mix of both scenarios above is possible:
+ * block1 = [sb11, sb12, sb13] // where sb13 is smaller than window length
+ * block2 = [b2]
+ *
+ * window.next() // sb11
+ * window.next() // sb12
+ * window.next() // [sb13, b2]
+ *
+ * @param blocks - sequence of blocks to manage
+ * @param windowSize - the size (in bytes) that block ranges should fit
+ * @tparam T - the specific type of `BlockWithSize`
+ * @note this class does not own `transferBlocks`
+ * @note this class is not thread safe
+ */
+class WindowedBlockIterator[T <: BlockWithSize](blocks: Seq[T], windowSize: Long)
+    extends Iterator[Seq[BlockRange[T]]] {
+
+  require(windowSize > 0, s"Invalid window size specified $windowSize")
+
+  case class BlockWindow(start: Long, size: Long) {
```

Comment:
```suggestion
  private case class BlockWindow(start: Long, size: Long) {
```

Reply:
This comment applies to the other case classes that are specific to this class.

Reply:
Agree. `rangeEnd` is actually quite private. In all other code external to the file I end up using `rangeSize`. I'll remove -1's, and +1's.

Will also fix the access of `BlockWindow`

Reply:
done.